File: src/php_implementation/sort/quick_sort/lomuto_partition.php
<?php

$start = microtime(true);

$array = [];
$array = randomArrayGenerator($array, 100_000);

function randomArrayGenerator(array $array, int $number) {
    for ($i = 0; $i < $number; $i++) {
        $array[] = rand(-1_000_000, 1_000_000);
    }

    return $array;
}

function quickSort(array &$array, $left, $right): void
{
    if ($left >= $right) {
        return;
    }

    $pivot = lomutoPartition($array, $left, $right);

    quickSort($array, $left, $pivot - 1);
    quickSort($array, $pivot + 1, $right);
}

function lomutoPartition(array &$array, int $left, int $right): int
{
    $pivot = $array[$right];
    $i = $left - 1;

    for ($j = $left; $j < $right; $j++) {
        if ($array[$j] <= $pivot) {
            $i++;
            $temp = $array[$i];
            $array[$i] = $array[$j];
            $array[$j] = $temp;
        }
    }

    $i++;
    $temp = $array[$i];
    $array[$i] = $array[$right];
    $array[$right] = $temp;

    return $i;
}

quickSort($array, 0, count($array) - 1);
print_r($array);

$end = microtime(true);
echo "Time: " . ($end - $start) . "s\n";

File: src/cpp_implementation/sort/quick_sort/lomuto_partition.cpp
#include <iostream>
#include <chrono>

using namespace std;
using namespace std::chrono;

void quickSort(int array[], int left, int right);
int lomutoPartition(int array[], int left, int right);

int main() {
    int array[100000];
    int n;
    cin >> n;
    srand(time(NULL));

    for (int i = 0; i < n; i++) {
        array[i] = rand() % 100000000;
    }

    auto start = high_resolution_clock::now();

    quickSort(array, 0, n - 1);
    for (int i = 0; i < n; i++) {
        cout << array[i] << "\n";
    }

    auto end = high_resolution_clock::now();
    auto duration = duration_cast<milliseconds>(end - start);

    cout << "Time: " << duration.count() << " ms" << endl;
}

void quickSort(int array[], int left, int right) {
    if (left >= right) {
        return;
    }

    int pivot = lomutoPartition(array, left, right);
    quickSort(array, left, pivot - 1);
    quickSort(array, pivot + 1, right);
}

int lomutoPartition(int array[], int left, int right) {
    int pivot = array[right];
    int i = left - 1;

    for (int j = left; j < right; j++) {
        if (array[j] <= pivot) {
            i++;
            swap(array[i], array[j]);
        }
    }

    // Swap pivot
    i++;
    swap(array[i], array[right]);

    return i; // return pivot position
}